Store fallback peers in TransactionFetcher as a PeerId, usize tuple

AbhinavMir opened the pull request:
potential fix #6528

emhane commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -870,7 +882,7 @@ pub(super) struct TxFetchMetadata {
     retries: u8,
     /// Peers that have announced the hash, but to which a request attempt has not yet been made.
-    fallback_peers: LruCache<PeerId>,
+    fallback_peers: LruCache<(PeerId, usize)>,
```
could you please make a new type
```rust
struct TxSizeMetadata {
    peer_id: PeerId
    tx_encoded_len: usize,
}
```

and then implement `Hash` and `Eq` for this, so that only the `PeerId` is considered. that way if a peer announces the same hash twice with two different sizes, it's only inserted once into the cache and the second time, the first entry is replaced.

AbhinavMir replied:
That makes more sense, I did consider doing it initially, will implement it by later today!

emhane replied:
we want to keep the cache with it's nice lru properties that suit our use case here, so a `LruCache<TxSizeMetada>`.

emhane replied:
although we could use an `LruMap`, and add the `PeerId` as key and `PackedOption<usize>` as the value. otherwise, we would need a method `TxSizeMetadata::get_size(&self) -> Option<usize>` which returns `None` if the size is zero. as mentioned in the issue, zero is a reserved value here.

AbhinavMir replied:
Hi @emhane , I saw this issue was closed. Should I reopen this? I wasn't working yesterday so didn't have a chance to get back.

Thanks!

emhane replied:
issue was still open, pr closed though but I reopened it :)

AbhinavMir replied:
Thanks! Yes, my bad, PR not issue. I have the changes ready. Should I build off of main or from my current branch? Thanks!

emhane replied:
off of main! so rebase your changes on main, shouldn't be too many conflicts since you already had `emhane/speed-up-request-buffered-hashes` as base which has now been merged into main

AbhinavMir replied:
There's actually some non-trivial changes I had to work with, so I will implement the changes on main directly!

AbhinavMir replied:
@emhane Is PackedOption from the Cranelift crate?

emhane replied:
yepp that's right! since you implemented `get_size` now manually, we don't need to use that type. we can just use the `TxSizeMetadata` type directly in the fallback peers cache, so we revert to a `LruCache` from `LruMap`, since we don't need a kv map anymore, just a set. We would have used `PackedOption<usize>` for the `tx_encoded_len` otherwise. `PackedOption` essentially does the same thing: has a reserved value (in our case 0), and stores that in place of the `Option` variant `None`. this saves memory, check out https://stackoverflow.com/a/16515488. try running that program in Rust Playground with `usize` (`tx_encoded_len` type).

AbhinavMir replied:
Makes perfect sense, did some playing around with it. I've made a decent amount of progress with using LruCache. Just wanted to ask: Is `peer_id` strictly unique? For the `fallback_peers.removal()` type functions, the `remove()` method expects value (as defined [here](https://github.com/AbhinavMir/reth/blob/f1a102ec38443af4db072de75d29c2c5d0a97316/crates/net/network/src/cache.rs#L60-L62)), so I can "get" the txSizeMetadata object, and pass it to the remove method, but for that I just wanted to be sure that `peer_id` is unique, and then find a way to implement `get_txSizeMetadata_from_peer_id()`. Thx!

emhane replied:
`peer_id` is unique!

emhane commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -870,7 +882,7 @@ pub(super) struct TxFetchMetadata {
-    fallback_peers: LruCache<PeerId>,
+    fallback_peers: LruCache<(PeerId, usize)>,
     /// Size metadata of the transaction if it has been seen in an eth68 announcement.
     // todo: store all seen sizes as a `(size, peer_id)` tuple to catch peers that respond with
     // another size tx than they announced. alt enter in request (won't catch peers announcing
```
```suggestion
```
this pr checks that to-do off, nice!

emhane commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -800,8 +812,8 @@ impl Stream for TransactionFetcher {
                 "`%peer_id` has been removed from `@active_peers` before inflight request(s) resolved, broken invariant `@active_peers` and `@inflight_requests`,
-`%peer_id`: {},
-`@self`: {:?}",
+    `%peer_id`: {},
+    `@self`: {:?}",
```
```suggestion
`%peer_id`: {},
`@self`: {:?}",
```
this way it's printed to stdout without indentation

AbhinavMir replied:
my bad! missed this!

emhane commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -281,26 +282,30 @@ impl TransactionFetcher {
             debug_assert!(
-                self.hashes_fetch_inflight_and_pending_fetch.peek(&hash).is_some(),
-                "`%hash` in `@buffered_hashes` that's not in `@unknown_hashes`, `@buffered_hashes` should be a subset of keys in `@unknown_hashes`, broken invariant `@buffered_hashes` and `@unknown_hashes`,
+            self.hashes_fetch_inflight_and_pending_fetch.peek(&hash).is_some(),
+            "`%hash` in `@buffered_hashes` that's not in `@unknown_hashes`, `@buffered_hashes` should be a subset of keys in `@unknown_hashes`, broken invariant `@buffered_hashes` and `@unknown_hashes`,
 `%hash`: {hash},
 `@self`: {self:?}",
-            );
+        );
```
please revert this formatting

emhane reviewed (changes requested):
coming along well

emhane commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -898,12 +898,24 @@ impl Hash for TxSizeMetadata {
     }
 }
 
+impl TxSizeMetadata {
+    fn get_size(&self) -> Option<usize> {
+        if self.tx_encoded_len == 0 {
+            None
+        } else {
+            Some(self.tx_encoded_len)
+        }
+    }
+}
+
+
```
looks good!

emhane commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -898,12 +898,24 @@ impl Hash for TxSizeMetadata {
     retries: u8,
     /// Peers that have announced the hash, but to which a request attempt has not yet been made.
+
     fallback_peers: LruCache<PeerId>,
```
some recent investigation has lead us to remove this dependency from the codebase...instead please use `LruMap<PeerId, ()>`

AbhinavMir replied:
Gotcha, will do.

emhane replied:
another pr is addressing this https://github.com/paradigmxyz/reth/pull/7312 by changing `LruCache` internally, so it's ok to reuse it here

emhane reviewed (commented):
nice progress!

emhane commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -879,13 +879,44 @@ impl Default for TransactionFetcher {
     /// Peers that have announced the hash, but to which a request attempt has not yet been made.
-    fallback_peers: LruCache<PeerId>,
+
+    fallback_peers: LruMap<PeerId, PackedOption<usize>>,
```
now that you made the type `TxSizeMetadata`, we don't need to import `PackedOption` :) the types are logically equivalent for our use case

emhane commented:
close to finish line @AbhinavMir !

onbjerg commented:
Hey @AbhinavMir! Anything we can help with to get this over the line? 😊 Let me know!

AbhinavMir commented:
Hi @onbjerg - I had lost context due to being busy, but wondering if I am back in the right direction? I removed packedOption and updated methods all around. There should be a warning, but if everything else looks good pls lmk!

emhane commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -899,13 +910,7 @@ impl Hash for TxSizeMetadata {
-        } else {
-            Some(self.tx_encoded_len)
-        }
-    }
```
this is the code that is equivalent to `PackedOption`. so removing this OR `PackedOption`, that was the goal. not both. sorry if this was too advanced, hope you learned smthg new :)

AbhinavMir replied:
Oops! I missed this, let me write in a fix!

AbhinavMir replied:
My bad! I missed this! I've reverted it. I am a little new to Rust was wasn't the easiest, but lots of fun learning.

emhane replied:
ok looks really good now!

AbhinavMir replied:
Thank you, let me know if I need anything else to have it merged. Unsure what should I do about the warnings generated during build, though.

emhane replied:
before a PR is merged, CI should be all green, have you tried fixing merge conflicts?

AbhinavMir replied:
Just fixed the merge conflicts, but it seems like some methods and variables, since newly introduced, are unused. What is best steps for dealing with this? I can add a few tests too.

emhane reviewed (commented):
maybe @onbjerg can better help you out from here, thanks for the effort @AbhinavMir !

AbhinavMir commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -98,8 +100,8 @@ impl TransactionFetcher {
-    pub(super) fn is_idle(&self, peer_id: &PeerId) -> bool {
-        let Some(inflight_count) = self.active_peers.peek(peer_id) else { return true };
+    pub(super) fn is_idle(&self, metadata: &TxSizeMetadata) -> bool {
+        let Some(inflight_count) = self.active_peers.peek(&metadata.peer_id) else { return true };
```
Done! Would this also change how the getter function for this (`get_idle_peer_for`) would work? I'll correspond with Oliver on this!

onbjerg commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -357,13 +357,15 @@ impl TransactionFetcher {
             if let Some(entry) = self.hashes_fetch_inflight_and_pending_fetch.get(hash) {
-                entry.fallback_peers_mut().remove(peer_failed_to_serve);
-                return true
+                entry.fallback_peers_mut().remove(&TxSizeMetadata {
+                    peer_id: *peer_failed_to_serve,
+                    tx_encoded_len: 0,
+                });
```
this doesn't seem correct (i see this over the rest of the pr too), will this not only remove it if the entry matches the peer id and `tx_encoded_len` is 0?

emhane replied:
nope, because of `PartialEq` impl, it only considers the `PeerId`. at best this would be abstracted away by adding a method to `TxFetchMetadata`
```rust
pub fn remove_fallback_peer(&mut self, peer_id: PeerId) {
    ...
}
```

onbjerg commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -1072,13 +1081,45 @@ impl Default for TransactionFetcher {
+
+impl Hash for TxSizeMetadata {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.peer_id.hash(state);
```
i think this is unexpected behavior, so not sure i am a fan. it should at least be documented

onbjerg commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -1072,13 +1081,45 @@ impl Default for TransactionFetcher {
+}
+
+impl TxSizeMetadata {
+    fn get_size(&self) -> Option<usize> {
+        if self.tx_encoded_len == 0 {
+            None
+        } else {
+            Some(self.tx_encoded_len)
+        }
+    }
```
any reason this doesn't just return `usize`?

emhane replied:
have you read the [docs](https://docs.rs/cranelift-entity/0.106.2/cranelift_entity/packed_option/index.html) for `PackedOption` type mentioned in the issue? it's similar to that type

onbjerg replied:
yes, i understand that it is the same as `PackedOption`, but i do not see why we need an option in the first place. is it because you want 0 to represent that the peer has not told us how big the tx is?

emhane commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -1072,13 +1081,45 @@ impl Default for TransactionFetcher {
 }
 
+#[derive(Debug, Clone)]
+pub(super) struct TxSizeMetadata {
```
```suggestion
pub struct TxSizeMetadata {
```

this fixes one warning, please address the other warnings about missing documentation. as for the warning about unused function name, please add an underscore at the start of the method name. I will use the methods when this pr is merged.

emhane commented on crates/net/network/src/transactions/fetcher.rs:
```diff
@@ -1072,13 +1081,45 @@ impl Default for TransactionFetcher {
     }
 }
 
+#[derive(Debug, Clone)]
```
```suggestion
#[derive(Debug, Clone, Constructor)]
```

one last thing to address, asides that which @onbjerg says, is to import `derive_more::Constructor`, then replace all the places in the code where you make an instance of the type. so this
```rust
TxSizeMetadat::new(peer_id, 0)
```
instead of making it like this 
```rust
TxSizeMetadata {
    peer_id: <value>,
    len: <value>,
}
```

emhane commented:
btw @AbhinavMir, try the command `make lint` from the Makefile

mattsse commented:
what to do with this @emhane ?

please take over or close

emhane commented:
> what to do with this @emhane ?
> 
> please take over or close

ok, I can take over :)
